src/mono/mono/utils/mono-wasm-pagemgr.h

```diff
@@ -0,0 +1,81 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+// Provides a simple, efficient implementation of anonymous mmap/munmap for WASM.
+// The emscripten libc version of mmap for anonymous mappings is slow during startup
+//  because it spends a lot of time zeroing already-zeroed pages, and in some cases
+//  the allocated pages will then get zeroed one more time by a malloc0 operation.
+
+// This is distinct from malloc/free in that we support freeing a subset of an allocation,
+//  i.e. you can allocate 4 pages and then free page 1, leaving pages 0/2/3 allocated.
+
+// Note that pages mapped by MWPM can't be unmapped by libc, and vice versa.
+
+#ifndef __MONO_UTILS_WASM_PAGEMGR_H__
+#define __MONO_UTILS_WASM_PAGEMGR_H__
+
+#include <stddef.h>
+#include <stdint.h>
+
+// 64KB (WASM native page size)
+// NOTE: emscripten sbrk() allows you to allocate less than 1 page, and doesn't
+//  return page-aligned addresses. So this value can theoretically be smaller.
+// Shrinking this would reduce wasted memory if callers try to mmap less than a
+//  whole page. My initial testing didn't show big improvements though, and it
+//  would make mmap (and to a lesser degree, munmap) have to scan/touch more pages.
+// I tested with 16KB pages and it wasn't an improvement over 64KB.
+// 8KB pages produces a slight reduction in total memory usage (1.3%). 4KB is worse.
+#define MWPM_PAGE_SIZE (64 * 1024)
+
+// 4GB (even though JS can't consistently handle addresses above 2GB)
+// System.Text.JSON.Tests needs to allocate more than 2GB...
+#define MWPM_MAX_MEMORY ((1024UL * 1024UL * 1024UL) * 4UL)
```

Comment:
@kg should we make some provision for wasm64? or is it obvious how to modify mwpm for wasm64 support?

Reply:
I think wasm64 would require an entirely different page table representation